error on my mac.

```
Undefined symbols for architecture x86_64:
  "llama::DataGenerator::generateRandomData(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >*, int)", referenced from:
      SortCorrectnessTest::SetUp() in SortUnitTest.cpp.o
  "llama::DataGenerator::generateSingleDatum(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, cases)", referenced from:
      SortObjectTest_EqualString_Test::TestBody() in SortUnitTest.cpp.o
      SortObjectTest_NotEqualString_Test::TestBody() in SortUnitTest.cpp.o
      SortObjectTest_GreaterString_Test::TestBody() in SortUnitTest.cpp.o
      SortObjectTest_LessString_Test::TestBody() in SortUnitTest.cpp.o
      DataGeneratorExceptionTest_GenerateSingleDatumUnrecognisedCase_Test::TestBody() in UtilityDataGeneratorUnitTest.cpp.o
  "llama::DataGenerator::generateContinuousData(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >*, int, bool)", referenced from:
      SortCorrectnessTest::SetUp() in SortUnitTest.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [/Users/sun/Documents/development/llama/bin_llama_cpp/Llama_UnitTest] Error 1
make[1]: *** [CMakeFiles/Llama_UnitTest.dir/all] Error 2
make: *** [all] Error 2
```

--- a/llama_cpp/src/Utility/DataGenerator.h
+++ b/llama_cpp/src/Utility/DataGenerator.h
@@ -186,25 +186,25 @@ class DataGenerator {
             }
             case LE: {
                 return DataGenerator::generateRandomDataFromRange<T>
-                (std::numeric_limits<T>::min(), givenData);
+                   (std::numeric_limits<T>::min(), givenData);
             }
             case GE: {
                 return DataGenerator::generateRandomDataFromRange<T>
-                (givenData, std::numeric_limits<T>::max());
+                   (givenData, std::numeric_limits<T>::max());
             }
             case LT: {
                 if (givenData == std::numeric_limits<T>::min()) {
-                    throw std::invalid_argument("The given value is the minimum of the spcific type.");
+                    throw std::invalid_argument("The given value is the minimum of the specified type.");
                 }
                 return DataGenerator::generateRandomDataFromRange<T>
-                (std::numeric_limits<T>::min(), givenData - std::numeric_limits<T>::denorm_min());
+                   (std::numeric_limits<T>::min(), givenData - std::numeric_limits<T>::denorm_min());
             }
             case GT: {
                 if (givenData == std::numeric_limits<T>::max()) {
-                    throw std::invalid_argument("The given value is the maximum of the spcific type.");
+                    throw std::invalid_argument("The given value is the maximum of the specified type.");
                 }
                 return DataGenerator::generateRandomDataFromRange<T>
-                (givenData + std::numeric_limits<T>::denorm_min(), std::numeric_limits<T>::max());
+                   (givenData + std::numeric_limits<T>::denorm_min(), std::numeric_limits<T>::max());
             }
             case EQ: {
                 return givenData;

--- a/llama_cpp/src/Utility/DataGenerator.h
+++ b/llama_cpp/src/Utility/DataGenerator.h
@@ -39,6 +39,9 @@
     #define DATA_GENERATOR_DEFAULT_STRING_MAX_LENGTH 20
 #endif
 
+#define TYPE_INTEGRAL typename std::enable_if<std::is_integral<T>::value, T>::type
+#define TYPE_FLOATING_POINT typename std::enable_if<std::is_floating_point<T>::value, T>::type
+
 typedef enum cases {LE, GE, LT, GT, EQ, NE, RD} Cases;
 
 namespace llama {
@@ -49,8 +52,22 @@ class DataGenerator {
     /**
      *  @brief Generate random data (arithmetic, aka, integer number and floating point number)
      */
-    template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
-    static void generateRandomData(std::vector<T> *targetVector, int count = DATA_GENERATOR_DEFAULT_COUNT) {
+    template<typename T, typename = typename std::enable_if<std::is_integral<T>::value, T>::type>
+    static void generateRandomData(std::vector<TYPE_INTEGRAL> *targetVector,
+                                   int count = DATA_GENERATOR_DEFAULT_COUNT) {
+        T typeMin = std::numeric_limits<T>::min();
+        T typeMax = std::numeric_limits<T>::max();
+        std::random_device arithmeticRandomDevice;
+        std::mt19937 arithmeticRandomEngine(arithmeticRandomDevice());
+        std::uniform_int_distribution<T> uniformArithmeticDistribution(typeMin, typeMax);
+
+        for (int i = 0; i < count; i++) {
+            targetVector->push_back(uniformArithmeticDistribution(arithmeticRandomEngine));
+        }
+    }
+    template<typename T, typename = typename std::enable_if<std::is_floating_point<T>::value, T>::type>
+    static void generateRandomData(std::vector<TYPE_FLOATING_POINT> *targetVector,
+                                   int count = DATA_GENERATOR_DEFAULT_COUNT) {
         T typeMin = std::numeric_limits<T>::min();
         T typeMax = std::numeric_limits<T>::max();
         std::random_device arithmeticRandomDevice;
@@ -208,10 +225,22 @@ class DataGenerator {
     /**
      *  @brief Generate single datum from a given range (arithmetic, aka, integer number and floating point number)
      */
-    template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
-    static T generateRandomDataFromRange(T min, T max) {
+    template<typename T>
+    static T generateRandomDataFromRange(TYPE_INTEGRAL min,
+                                         TYPE_INTEGRAL max) {
         std::random_device randomDevice;
         std::mt19937 randomEngine(randomDevice());
+
+        std::uniform_int_distribution<T> uniformDistribution(min, max);
+        return uniformDistribution(randomEngine);
+    }
+
+    template<typename T>
+    static T generateRandomDataFromRange(TYPE_FLOATING_POINT min,
+                                         TYPE_FLOATING_POINT max) {
+        std::random_device randomDevice;
+        std::mt19937 randomEngine(randomDevice());
+
         std::uniform_real_distribution<T> uniformDistribution(min, max);
         return uniformDistribution(randomEngine);
     }

--- a/llama_cpp/src/Utility/DataGenerator.h
+++ b/llama_cpp/src/Utility/DataGenerator.h
@@ -50,11 +50,12 @@ class DataGenerator {
     DataGenerator() {}
 
     /**
-     *  @brief Generate random data (arithmetic, aka, integer number and floating point number)
+     *  @brief Generate random integral numbers
      */
-    template<typename T, typename = typename std::enable_if<std::is_integral<T>::value, T>::type>
-    static void generateRandomData(std::vector<TYPE_INTEGRAL> *targetVector,
-                                   int count = DATA_GENERATOR_DEFAULT_COUNT) {
+    template<typename T>
+    static void generateRandomData(std::vector<T> *targetVector,
+                                   int count = DATA_GENERATOR_DEFAULT_COUNT,
+                                   TYPE_INTEGRAL* = 0) {
         T typeMin = std::numeric_limits<T>::min();
         T typeMax = std::numeric_limits<T>::max();
         std::random_device arithmeticRandomDevice;
@@ -65,9 +66,14 @@ class DataGenerator {
             targetVector->push_back(uniformArithmeticDistribution(arithmeticRandomEngine));
         }
     }
-    template<typename T, typename = typename std::enable_if<std::is_floating_point<T>::value, T>::type>
-    static void generateRandomData(std::vector<TYPE_FLOATING_POINT> *targetVector,
-                                   int count = DATA_GENERATOR_DEFAULT_COUNT) {
+
+    /**
+     *  @brief Generate random floating point numbers
+     */
+    template<typename T>
+    static void generateRandomData(std::vector<T> *targetVector,
+                                   int count = DATA_GENERATOR_DEFAULT_COUNT,
+                                   TYPE_FLOATING_POINT* = 0) {
         T typeMin = std::numeric_limits<T>::min();
         T typeMax = std::numeric_limits<T>::max();
         std::random_device arithmeticRandomDevice;
@@ -226,8 +232,8 @@ class DataGenerator {
      *  @brief Generate single datum from a given range (arithmetic, aka, integer number and floating point number)
      */
     template<typename T>
-    static T generateRandomDataFromRange(TYPE_INTEGRAL min,
-                                         TYPE_INTEGRAL max) {
+    static T generateRandomDataFromRange(T min, T max,
+                                         TYPE_INTEGRAL* = 0) {
         std::random_device randomDevice;
         std::mt19937 randomEngine(randomDevice());
 
@@ -236,8 +242,8 @@ class DataGenerator {
     }
 
     template<typename T>
-    static T generateRandomDataFromRange(TYPE_FLOATING_POINT min,
-                                         TYPE_FLOATING_POINT max) {
+    static T generateRandomDataFromRange(T min, T max,
+                                         TYPE_FLOATING_POINT* = 0) {
         std::random_device randomDevice;
         std::mt19937 randomEngine(randomDevice());
 

--- a/llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp
+++ b/llama_cpp/test/Llama_UnitTest/SortUnitTest.cpp
@@ -229,14 +229,14 @@ struct SortCorrectnessTest : public ::testing::Test {
     llama::SelectionSort<string> stringSe;
 
     void SetUp() override {
-        DataGenerator::generateRandomData(&randomInts, 10);
-        DataGenerator::generateRandomData(&randomDoubles, 10);
-        DataGenerator::generateRandomData(&randomStrings, 10);
+        DataGenerator::generateRandomData(&randomInts);
+        DataGenerator::generateRandomData(&randomDoubles);
+        DataGenerator::generateRandomData(&randomStrings);
         DataGenerator::generateContinuousData(&continuousInts);
         DataGenerator::generateContinuousData(&continuousDoubles);
-        DataGenerator::generateContinuousData(&continuousStrings, 10, false);
-        DataGenerator::generateContinuousData(&continuousIntsReversed);
-        DataGenerator::generateContinuousData(&continuousDoublesReversed);
+        DataGenerator::generateContinuousData(&continuousStrings);
+        DataGenerator::generateContinuousData(&continuousIntsReversed, 0, 1, 10, true);
+        DataGenerator::generateContinuousData(&continuousDoublesReversed, 0.0, 0.1, 10, true);
         DataGenerator::generateContinuousData(&continuousStringsReversed, 10, true);
 
         randomIntsCorrect = randomInts;

--- a/llama_cpp/test/Llama_UnitTest/UtilityDataGeneratorUnitTest.cpp
+++ b/llama_cpp/test/Llama_UnitTest/UtilityDataGeneratorUnitTest.cpp
@@ -53,7 +53,7 @@ TEST(DataGeneratorExceptionTest, GenerateSingleDatumMinimumInteger) {
     try {
         DataGenerator::generateSingleDatum(minimumInt, LT);
     } catch(std::invalid_argument const & err) {
-        EXPECT_EQ(err.what(), std::string("The given value is the minimum of the spcific type."));
+        EXPECT_EQ(err.what(), std::string("The given value is the minimum of the specified type."));
     } catch(...) {
         FAIL() << "Expected std::invalid_argument";
     }
@@ -73,7 +73,7 @@ TEST(DataGeneratorExceptionTest, GenerateSingleDatumMaximumInteger) {
     try {
         DataGenerator::generateSingleDatum(maximumInt, GT);
     } catch(std::invalid_argument const & err) {
-        EXPECT_EQ(err.what(), std::string("The given value is the maximum of the spcific type."));
+        EXPECT_EQ(err.what(), std::string("The given value is the maximum of the specified type."));
     } catch(...) {
         FAIL() << "Expected std::invalid_argument";
     }
@@ -93,7 +93,7 @@ TEST(DataGeneratorExceptionTest, GenerateSingleDatumMinimumDouble) {
     try {
         DataGenerator::generateSingleDatum(minimumDouble, LT);
     } catch(std::invalid_argument const & err) {
-        EXPECT_EQ(err.what(), std::string("The given value is the minimum of the spcific type."));
+        EXPECT_EQ(err.what(), std::string("The given value is the minimum of the specified type."));
     } catch(...) {
         FAIL() << "Expected std::invalid_argument";
     }
@@ -113,7 +113,7 @@ TEST(DataGeneratorExceptionTest, GenerateSingleDatumMaximumDouble) {
     try {
         DataGenerator::generateSingleDatum(maximumDouble, GT);
     } catch(std::invalid_argument const & err) {
-        EXPECT_EQ(err.what(), std::string("The given value is the maximum of the spcific type."));
+        EXPECT_EQ(err.what(), std::string("The given value is the maximum of the specified type."));
     } catch(...) {
         FAIL() << "Expected std::invalid_argument";
     }

